File: auth-authorization-server-api/src/main/java/com/cartobucket/auth/authorization/server/dto/WellKnown.java
package com.cartobucket.auth.authorization.server.dto;

import jakarta.json.bind.annotation.JsonbProperty;
import jakarta.json.bind.annotation.JsonbNillable;
import jakarta.json.bind.config.PropertyNamingStrategy;
import jakarta.validation.Valid;
import java.util.List;
import java.util.Objects;

@JsonbNillable(false)
public class WellKnown {
    
    public enum ResponseTypesSupportedEnum {
        CODE("code"), 
        TOKEN("token"), 
        CODE_ID_TOKEN("code id_token");

        private String value;

        ResponseTypesSupportedEnum(String v) {
            value = v;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static ResponseTypesSupportedEnum fromString(String s) {
            for (ResponseTypesSupportedEnum b : ResponseTypesSupportedEnum.values()) {
                if (java.util.Objects.toString(b.value).equals(s)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected string value '" + s + "'");
        }

        public static ResponseTypesSupportedEnum fromValue(String value) {
            for (ResponseTypesSupportedEnum b : ResponseTypesSupportedEnum.values()) {
                if (b.value.equals(value)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected value '" + value + "'");
        }
    }

    public enum SubjectTypesSupportedEnum {
        PUBLIC("public");

        private String value;

        SubjectTypesSupportedEnum(String v) {
            value = v;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static SubjectTypesSupportedEnum fromString(String s) {
            for (SubjectTypesSupportedEnum b : SubjectTypesSupportedEnum.values()) {
                if (java.util.Objects.toString(b.value).equals(s)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected string value '" + s + "'");
        }

        public static SubjectTypesSupportedEnum fromValue(String value) {
            for (SubjectTypesSupportedEnum b : SubjectTypesSupportedEnum.values()) {
                if (b.value.equals(value)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected value '" + value + "'");
        }
    }

    public enum IdTokenSigningAlgValuesSupportedEnum {
        RS256("RS256"), 
        RS512("RS512"), 
        EDDSA("EdDSA");

        private String value;

        IdTokenSigningAlgValuesSupportedEnum(String v) {
            value = v;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static IdTokenSigningAlgValuesSupportedEnum fromString(String s) {
            for (IdTokenSigningAlgValuesSupportedEnum b : IdTokenSigningAlgValuesSupportedEnum.values()) {
                if (java.util.Objects.toString(b.value).equals(s)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected string value '" + s + "'");
        }

        public static IdTokenSigningAlgValuesSupportedEnum fromValue(String value) {
            for (IdTokenSigningAlgValuesSupportedEnum b : IdTokenSigningAlgValuesSupportedEnum.values()) {
                if (b.value.equals(value)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected value '" + value + "'");
        }
    }

    public enum ScopesSupportedEnum {
        OPENID("openid"), 
        EMAIL("email"), 
        PROFILE("profile");

        private String value;

        ScopesSupportedEnum(String v) {
            value = v;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static ScopesSupportedEnum fromString(String s) {
            for (ScopesSupportedEnum b : ScopesSupportedEnum.values()) {
                if (java.util.Objects.toString(b.value).equals(s)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected string value '" + s + "'");
        }

        public static ScopesSupportedEnum fromValue(String value) {
            for (ScopesSupportedEnum b : ScopesSupportedEnum.values()) {
                if (b.value.equals(value)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected value '" + value + "'");
        }
    }

    public enum TokenEndpointAuthMethodsSupportedEnum {
        POST("client_secret_post"), 
        BASIC("client_secret_basic");

        private String value;

        TokenEndpointAuthMethodsSupportedEnum(String v) {
            value = v;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static TokenEndpointAuthMethodsSupportedEnum fromString(String s) {
            for (TokenEndpointAuthMethodsSupportedEnum b : TokenEndpointAuthMethodsSupportedEnum.values()) {
                if (java.util.Objects.toString(b.value).equals(s)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected string value '" + s + "'");
        }

        public static TokenEndpointAuthMethodsSupportedEnum fromValue(String value) {
            for (TokenEndpointAuthMethodsSupportedEnum b : TokenEndpointAuthMethodsSupportedEnum.values()) {
                if (b.value.equals(value)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected value '" + value + "'");
        }
    }

    public enum ClaimsSupportedEnum {
        AUD("aud"), 
        EXP("exp"), 
        FAMILY_NAME("family_name"), 
        GIVEN_NAME("given_name"), 
        IAT("iat"), 
        ISS("iss"), 
        NAME("name"), 
        SUB("sub");

        private String value;

        ClaimsSupportedEnum(String v) {
            value = v;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static ClaimsSupportedEnum fromString(String s) {
            for (ClaimsSupportedEnum b : ClaimsSupportedEnum.values()) {
                if (java.util.Objects.toString(b.value).equals(s)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected string value '" + s + "'");
        }

        public static ClaimsSupportedEnum fromValue(String value) {
            for (ClaimsSupportedEnum b : ClaimsSupportedEnum.values()) {
                if (b.value.equals(value)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected value '" + value + "'");
        }
    }

    public enum CodeChallengeMethodsSupportedEnum {
        S256("S256");

        private String value;

        CodeChallengeMethodsSupportedEnum(String v) {
            value = v;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static CodeChallengeMethodsSupportedEnum fromString(String s) {
            for (CodeChallengeMethodsSupportedEnum b : CodeChallengeMethodsSupportedEnum.values()) {
                if (java.util.Objects.toString(b.value).equals(s)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected string value '" + s + "'");
        }

        public static CodeChallengeMethodsSupportedEnum fromValue(String value) {
            for (CodeChallengeMethodsSupportedEnum b : CodeChallengeMethodsSupportedEnum.values()) {
                if (b.value.equals(value)) {
                    return b;
                }
            }
            throw new IllegalArgumentException("Unexpected value '" + value + "'");
        }
    }
    
    @JsonbProperty("issuer")
    private @Valid String issuer;
    
    @JsonbProperty("authorization_endpoint")
    private @Valid String authorizationEndpoint;
    
    @JsonbProperty("token_endpoint")
    private @Valid String tokenEndpoint;
    
    @JsonbProperty("userinfo_endpoint")
    private @Valid String userinfoEndpoint;
    
    @JsonbProperty("jwks_uri")
    private @Valid String jwksUri;
    
    @JsonbProperty("revocation_endpoint")
    private @Valid String revocationEndpoint;
    
    @JsonbProperty("scopes_supported")
    private @Valid List<String> scopesSupported;
    
    @JsonbProperty("response_types_supported")
    private @Valid List<String> responseTypesSupported;
    
    @JsonbProperty("grant_types_supported")
    private @Valid List<String> grantTypesSupported;
    
    @JsonbProperty("subject_types_supported")
    private @Valid List<String> subjectTypesSupported;
    
    @JsonbProperty("id_token_signing_alg_values_supported")
    private @Valid List<String> idTokenSigningAlgValuesSupported;
    
    @JsonbProperty("token_endpoint_auth_methods_supported")
    private @Valid List<String> tokenEndpointAuthMethodsSupported;
    
    @JsonbProperty("claims_supported")
    private @Valid List<String> claimsSupported;
    
    @JsonbProperty("code_challenge_methods_supported")
    private @Valid List<String> codeChallengeMethodsSupported;
    
    public WellKnown issuer(String issuer) {
        this.issuer = issuer;
        return this;
    }
    
    public String getIssuer() {
        return issuer;
    }
    
    public void setIssuer(String issuer) {
        this.issuer = issuer;
    }
    
    public WellKnown authorizationEndpoint(String authorizationEndpoint) {
        this.authorizationEndpoint = authorizationEndpoint;
        return this;
    }
    
    public String getAuthorizationEndpoint() {
        return authorizationEndpoint;
    }
    
    public void setAuthorizationEndpoint(String authorizationEndpoint) {
        this.authorizationEndpoint = authorizationEndpoint;
    }
    
    public WellKnown tokenEndpoint(String tokenEndpoint) {
        this.tokenEndpoint = tokenEndpoint;
        return this;
    }
    
    public String getTokenEndpoint() {
        return tokenEndpoint;
    }
    
    public void setTokenEndpoint(String tokenEndpoint) {
        this.tokenEndpoint = tokenEndpoint;
    }
    
    public WellKnown userinfoEndpoint(String userinfoEndpoint) {
        this.userinfoEndpoint = userinfoEndpoint;
        return this;
    }
    
    public String getUserinfoEndpoint() {
        return userinfoEndpoint;
    }
    
    public void setUserinfoEndpoint(String userinfoEndpoint) {
        this.userinfoEndpoint = userinfoEndpoint;
    }
    
    public WellKnown jwksUri(String jwksUri) {
        this.jwksUri = jwksUri;
        return this;
    }
    
    public String getJwksUri() {
        return jwksUri;
    }
    
    public void setJwksUri(String jwksUri) {
        this.jwksUri = jwksUri;
    }
    
    public WellKnown revocationEndpoint(String revocationEndpoint) {
        this.revocationEndpoint = revocationEndpoint;
        return this;
    }
    
    public String getRevocationEndpoint() {
        return revocationEndpoint;
    }
    
    public void setRevocationEndpoint(String revocationEndpoint) {
        this.revocationEndpoint = revocationEndpoint;
    }
    
    public WellKnown scopesSupported(List<String> scopesSupported) {
        this.scopesSupported = scopesSupported;
        return this;
    }
    
    public List<String> getScopesSupported() {
        return scopesSupported;
    }
    
    public void setScopesSupported(List<String> scopesSupported) {
        this.scopesSupported = scopesSupported;
    }
    
    public WellKnown responseTypesSupported(List<String> responseTypesSupported) {
        this.responseTypesSupported = responseTypesSupported;
        return this;
    }
    
    public List<String> getResponseTypesSupported() {
        return responseTypesSupported;
    }
    
    public void setResponseTypesSupported(List<String> responseTypesSupported) {
        this.responseTypesSupported = responseTypesSupported;
    }
    
    public WellKnown grantTypesSupported(List<String> grantTypesSupported) {
        this.grantTypesSupported = grantTypesSupported;
        return this;
    }
    
    public List<String> getGrantTypesSupported() {
        return grantTypesSupported;
    }
    
    public void setGrantTypesSupported(List<String> grantTypesSupported) {
        this.grantTypesSupported = grantTypesSupported;
    }
    
    public WellKnown subjectTypesSupported(List<String> subjectTypesSupported) {
        this.subjectTypesSupported = subjectTypesSupported;
        return this;
    }
    
    public List<String> getSubjectTypesSupported() {
        return subjectTypesSupported;
    }
    
    public void setSubjectTypesSupported(List<String> subjectTypesSupported) {
        this.subjectTypesSupported = subjectTypesSupported;
    }
    
    public WellKnown idTokenSigningAlgValuesSupported(List<String> idTokenSigningAlgValuesSupported) {
        this.idTokenSigningAlgValuesSupported = idTokenSigningAlgValuesSupported;
        return this;
    }
    
    public List<String> getIdTokenSigningAlgValuesSupported() {
        return idTokenSigningAlgValuesSupported;
    }
    
    public void setIdTokenSigningAlgValuesSupported(List<String> idTokenSigningAlgValuesSupported) {
        this.idTokenSigningAlgValuesSupported = idTokenSigningAlgValuesSupported;
    }
    
    public WellKnown tokenEndpointAuthMethodsSupported(List<String> tokenEndpointAuthMethodsSupported) {
        this.tokenEndpointAuthMethodsSupported = tokenEndpointAuthMethodsSupported;
        return this;
    }
    
    public List<String> getTokenEndpointAuthMethodsSupported() {
        return tokenEndpointAuthMethodsSupported;
    }
    
    public void setTokenEndpointAuthMethodsSupported(List<String> tokenEndpointAuthMethodsSupported) {
        this.tokenEndpointAuthMethodsSupported = tokenEndpointAuthMethodsSupported;
    }
    
    public WellKnown claimsSupported(List<String> claimsSupported) {
        this.claimsSupported = claimsSupported;
        return this;
    }
    
    public List<String> getClaimsSupported() {
        return claimsSupported;
    }
    
    public void setClaimsSupported(List<String> claimsSupported) {
        this.claimsSupported = claimsSupported;
    }
    
    public WellKnown codeChallengeMethodsSupported(List<String> codeChallengeMethodsSupported) {
        this.codeChallengeMethodsSupported = codeChallengeMethodsSupported;
        return this;
    }
    
    public List<String> getCodeChallengeMethodsSupported() {
        return codeChallengeMethodsSupported;
    }
    
    public void setCodeChallengeMethodsSupported(List<String> codeChallengeMethodsSupported) {
        this.codeChallengeMethodsSupported = codeChallengeMethodsSupported;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        WellKnown wellKnown = (WellKnown) o;
        return Objects.equals(this.issuer, wellKnown.issuer) &&
            Objects.equals(this.authorizationEndpoint, wellKnown.authorizationEndpoint) &&
            Objects.equals(this.tokenEndpoint, wellKnown.tokenEndpoint) &&
            Objects.equals(this.userinfoEndpoint, wellKnown.userinfoEndpoint) &&
            Objects.equals(this.jwksUri, wellKnown.jwksUri) &&
            Objects.equals(this.revocationEndpoint, wellKnown.revocationEndpoint) &&
            Objects.equals(this.scopesSupported, wellKnown.scopesSupported) &&
            Objects.equals(this.responseTypesSupported, wellKnown.responseTypesSupported) &&
            Objects.equals(this.grantTypesSupported, wellKnown.grantTypesSupported) &&
            Objects.equals(this.subjectTypesSupported, wellKnown.subjectTypesSupported) &&
            Objects.equals(this.idTokenSigningAlgValuesSupported, wellKnown.idTokenSigningAlgValuesSupported) &&
            Objects.equals(this.tokenEndpointAuthMethodsSupported, wellKnown.tokenEndpointAuthMethodsSupported) &&
            Objects.equals(this.claimsSupported, wellKnown.claimsSupported) &&
            Objects.equals(this.codeChallengeMethodsSupported, wellKnown.codeChallengeMethodsSupported);
    }

    @Override
    public int hashCode() {
        return Objects.hash(issuer, authorizationEndpoint, tokenEndpoint, userinfoEndpoint, jwksUri, revocationEndpoint, scopesSupported, responseTypesSupported, grantTypesSupported, subjectTypesSupported, idTokenSigningAlgValuesSupported, tokenEndpointAuthMethodsSupported, claimsSupported, codeChallengeMethodsSupported);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("class WellKnown {\n");
        sb.append("    issuer: ").append(toIndentedString(issuer)).append("\n");
        sb.append("    authorizationEndpoint: ").append(toIndentedString(authorizationEndpoint)).append("\n");
        sb.append("    tokenEndpoint: ").append(toIndentedString(tokenEndpoint)).append("\n");
        sb.append("    userinfoEndpoint: ").append(toIndentedString(userinfoEndpoint)).append("\n");
        sb.append("    jwksUri: ").append(toIndentedString(jwksUri)).append("\n");
        sb.append("    revocationEndpoint: ").append(toIndentedString(revocationEndpoint)).append("\n");
        sb.append("    scopesSupported: ").append(toIndentedString(scopesSupported)).append("\n");
        sb.append("    responseTypesSupported: ").append(toIndentedString(responseTypesSupported)).append("\n");
        sb.append("    grantTypesSupported: ").append(toIndentedString(grantTypesSupported)).append("\n");
        sb.append("    subjectTypesSupported: ").append(toIndentedString(subjectTypesSupported)).append("\n");
        sb.append("    idTokenSigningAlgValuesSupported: ").append(toIndentedString(idTokenSigningAlgValuesSupported)).append("\n");
        sb.append("    tokenEndpointAuthMethodsSupported: ").append(toIndentedString(tokenEndpointAuthMethodsSupported)).append("\n");
        sb.append("    claimsSupported: ").append(toIndentedString(claimsSupported)).append("\n");
        sb.append("    codeChallengeMethodsSupported: ").append(toIndentedString(codeChallengeMethodsSupported)).append("\n");
        sb.append("}");
        return sb.toString();
    }

    private String toIndentedString(Object o) {
        if (o == null) {
            return "null";
        }
        return o.toString().replace("\n", "\n    ");
    }
}

File: auth-authorization-server-api/src/main/kotlin/com/cartobucket/auth/authorization/server/dto/WellKnown.kt
/*
 * Copyright 2023 Bryce Groff (Cartobucket LLC)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.cartobucket.auth.authorization.server.dto

import jakarta.json.bind.annotation.JsonbNillable
import jakarta.json.bind.annotation.JsonbProperty

@JsonbNillable(false)
data class WellKnown(
    @field:JsonbProperty("issuer")
    var issuer: String? = null,

    @field:JsonbProperty("authorization_endpoint")
    var authorizationEndpoint: String? = null,

    @field:JsonbProperty("token_endpoint")
    var tokenEndpoint: String? = null,

    @field:JsonbProperty("userinfo_endpoint")
    var userinfoEndpoint: String? = null,

    @field:JsonbProperty("jwks_uri")
    var jwksUri: String? = null,

    @field:JsonbProperty("revocation_endpoint")
    var revocationEndpoint: String? = null,

    @field:JsonbProperty("scopes_supported")
    var scopesSupported: List<String>? = null,

    @field:JsonbProperty("response_types_supported")
    var responseTypesSupported: List<String>? = null,

    @field:JsonbProperty("grant_types_supported")
    var grantTypesSupported: List<String>? = null,

    @field:JsonbProperty("subject_types_supported")
    var subjectTypesSupported: List<String>? = null,

    @field:JsonbProperty("id_token_signing_alg_values_supported")
    var idTokenSigningAlgValuesSupported: List<String>? = null,

    @field:JsonbProperty("token_endpoint_auth_methods_supported")
    var tokenEndpointAuthMethodsSupported: List<String>? = null,

    @field:JsonbProperty("claims_supported")
    var claimsSupported: List<String>? = null,

    @field:JsonbProperty("code_challenge_methods_supported")
    var codeChallengeMethodsSupported: List<String>? = null
) {
    enum class ResponseTypesSupportedEnum(private val _value: String) {
        CODE("code"),
        TOKEN("token"),
        CODE_ID_TOKEN("code id_token");

        fun value(): String = _value

        override fun toString(): String = _value

        companion object {
            @JvmStatic
            fun fromString(s: String): ResponseTypesSupportedEnum =
                entries.find { it._value == s }
                    ?: throw IllegalArgumentException("Unexpected string value '$s'")

            @JvmStatic
            fun fromValue(value: String): ResponseTypesSupportedEnum =
                entries.find { it._value == value }
                    ?: throw IllegalArgumentException("Unexpected value '$value'")
        }
    }

    enum class SubjectTypesSupportedEnum(private val _value: String) {
        PUBLIC("public");

        fun value(): String = _value

        override fun toString(): String = _value

        companion object {
            @JvmStatic
            fun fromString(s: String): SubjectTypesSupportedEnum =
                entries.find { it._value == s }
                    ?: throw IllegalArgumentException("Unexpected string value '$s'")

            @JvmStatic
            fun fromValue(value: String): SubjectTypesSupportedEnum =
                entries.find { it._value == value }
                    ?: throw IllegalArgumentException("Unexpected value '$value'")
        }
    }

    enum class IdTokenSigningAlgValuesSupportedEnum(private val _value: String) {
        RS256("RS256"),
        RS512("RS512"),
        EDDSA("EdDSA");

        fun value(): String = _value

        override fun toString(): String = _value

        companion object {
            @JvmStatic
            fun fromString(s: String): IdTokenSigningAlgValuesSupportedEnum =
                entries.find { it._value == s }
                    ?: throw IllegalArgumentException("Unexpected string value '$s'")

            @JvmStatic
            fun fromValue(value: String): IdTokenSigningAlgValuesSupportedEnum =
                entries.find { it._value == value }
                    ?: throw IllegalArgumentException("Unexpected value '$value'")
        }
    }

    enum class ScopesSupportedEnum(private val _value: String) {
        OPENID("openid"),
        EMAIL("email"),
        PROFILE("profile");

        fun value(): String = _value

        override fun toString(): String = _value

        companion object {
            @JvmStatic
            fun fromString(s: String): ScopesSupportedEnum =
                entries.find { it._value == s }
                    ?: throw IllegalArgumentException("Unexpected string value '$s'")

            @JvmStatic
            fun fromValue(value: String): ScopesSupportedEnum =
                entries.find { it._value == value }
                    ?: throw IllegalArgumentException("Unexpected value '$value'")
        }
    }

    enum class TokenEndpointAuthMethodsSupportedEnum(private val _value: String) {
        POST("client_secret_post"),
        BASIC("client_secret_basic");

        fun value(): String = _value

        override fun toString(): String = _value

        companion object {
            @JvmStatic
            fun fromString(s: String): TokenEndpointAuthMethodsSupportedEnum =
                entries.find { it._value == s }
                    ?: throw IllegalArgumentException("Unexpected string value '$s'")

            @JvmStatic
            fun fromValue(value: String): TokenEndpointAuthMethodsSupportedEnum =
                entries.find { it._value == value }
                    ?: throw IllegalArgumentException("Unexpected value '$value'")
        }
    }

    enum class ClaimsSupportedEnum(private val _value: String) {
        AUD("aud"),
        EXP("exp"),
        FAMILY_NAME("family_name"),
        GIVEN_NAME("given_name"),
        IAT("iat"),
        ISS("iss"),
        NAME("name"),
        SUB("sub");

        fun value(): String = _value

        override fun toString(): String = _value

        companion object {
            @JvmStatic
            fun fromString(s: String): ClaimsSupportedEnum =
                entries.find { it._value == s }
                    ?: throw IllegalArgumentException("Unexpected string value '$s'")

            @JvmStatic
            fun fromValue(value: String): ClaimsSupportedEnum =
                entries.find { it._value == value }
                    ?: throw IllegalArgumentException("Unexpected value '$value'")
        }
    }

    enum class CodeChallengeMethodsSupportedEnum(private val _value: String) {
        S256("S256");

        fun value(): String = _value

        override fun toString(): String = _value

        companion object {
            @JvmStatic
            fun fromString(s: String): CodeChallengeMethodsSupportedEnum =
                entries.find { it._value == s }
                    ?: throw IllegalArgumentException("Unexpected string value '$s'")

            @JvmStatic
            fun fromValue(value: String): CodeChallengeMethodsSupportedEnum =
                entries.find { it._value == value }
                    ?: throw IllegalArgumentException("Unexpected value '$value'")
        }
    }
}
